[3/N][dtensor] Strided Sharding offset calculation util

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130760
* #131408
* __->__ #132391
* #130239
* #126697

**Summary**
1. change `compute_local_shape_and_global_offset` to correctly compute shape and offset for strided sharding placement (currently it only handles 2D and some 3D+ sharding).
2. Add a new property `num_shards_map` to `DTensorSpec` denoting how many shards each tensor dimension has. This is necessary for constructing `_StridedShard` placement when we call `distribute_tensor(dtensor_tp, dp_device_mesh, [Shard(0)])` and the `split_factor` argument will just be the number of shards on that sharding tensor dim.

**Test**
`test/distributed/_tensor/test_utils.py`

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

XilunWu committed:
[3/N][dtensor] Strided Sharding offset calculation util

[ghstack-poisoned]

XilunWu committed:
Update on "[3/N][dtensor] Strided Sharding offset calculation util"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60606115).

wanchaol commented on test/distributed/_tensor/test_utils.py:
```diff
@@ -127,6 +127,30 @@ def test_compute_local_shape_and_global_offset_2D(self):
+        )
+        # local shard shape is [2, 2]
+        global_tensor_shape = torch.Size([2 * self.world_size, 2])
+        placements = [_StridedShard(0, split_factor=tp_size), Shard(0)]
```
Can you also add a test case of [Replicate(), _StridedShard, Shard], this would cover HSDP state dict offset calculation

XilunWu committed:
Update on "[3/N][dtensor] Strided Sharding offset calculation util"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

Differential Revision: [D60606115](https://our.internmc.facebook.com/intern/diff/D60606115)

[ghstack-poisoned]

XilunWu committed:
Update on "[3/N][dtensor] Strided Sharding offset calculation util"


**Summary**
1. change `compute_local_shape_and_global_offset` to correctly compute shape and offset for strided sharding placement (currently it only handles 2D and some 3D+ sharding).
2. Add a new property `num_shards_map` to `DTensorSpec` denoting how many shards each tensor dimension has. This is necessary for constructing `_StridedShard` placement when we call `distribute_tensor(dtensor_tp, dp_device_mesh, [Shard(0)])` and the `split_factor` argument will just be the number of shards on that sharding tensor dim.

**Test**
`test/distributed/_tensor/test_utils.py`

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

Differential Revision: [D60606115](https://our.internmc.facebook.com/intern/diff/D60606115)

[ghstack-poisoned]

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60606115).

XilunWu committed:
Update on "[3/N][dtensor] Strided Sharding offset calculation util"


**Summary**
1. change `compute_local_shape_and_global_offset` to correctly compute shape and offset for strided sharding placement (currently it only handles 2D and some 3D+ sharding).
2. Add a new property `num_shards_map` to `DTensorSpec` denoting how many shards each tensor dimension has. This is necessary for constructing `_StridedShard` placement when we call `distribute_tensor(dtensor_tp, dp_device_mesh, [Shard(0)])` and the `split_factor` argument will just be the number of shards on that sharding tensor dim.

**Test**
`test/distributed/_tensor/test_utils.py`

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

Differential Revision: [D60606115](https://our.internmc.facebook.com/intern/diff/D60606115)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/_utils.py:
```diff
@@ -121,6 +126,58 @@ def compute_local_shape_and_global_offset(
+        # TODO: change this function to correctly address this.
+        # TODO: this logic can be applied to contiguous sharding as well
+        strided_sharding = any(isinstance(p, _StridedShard) for p in placements)
+        if strided_sharding:
```
Based on the note above, in this PR, can you throw a error inside this `if` branch to ensure when such a 3D sharding happen, we properly error out to users, and add a test to guard this

wanchaol reviewed (approved):
lgtm, have one more comment, please address before landing

XilunWu committed:
Update on "[3/N][dtensor] Strided Sharding offset calculation util"


**Summary**
1. change `compute_local_shape_and_global_offset` to correctly compute shape and offset for strided sharding placement (currently it only handles 2D and some 3D+ sharding).
2. Add a new property `num_shards_map` to `DTensorSpec` denoting how many shards each tensor dimension has. This is necessary for constructing `_StridedShard` placement when we call `distribute_tensor(dtensor_tp, dp_device_mesh, [Shard(0)])` and the `split_factor` argument will just be the number of shards on that sharding tensor dim.

**Test**
`test/distributed/_tensor/test_utils.py`

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

Differential Revision: [D60606115](https://our.internmc.facebook.com/intern/diff/D60606115)

[ghstack-poisoned]
